File: 2015/day06.rb
def part_one(instructions)
    grid = Array.new(1000) {Array.new(1000) {false}}

    instructions.each do | instruction |
        words = instruction.split()

        first, second = words[-3], words[-1]
        x_from, y_from = first.split(",").map(&:to_i)
        x_to, y_to = second.split(",").map(&:to_i)

        for x in x_from..x_to do
            for y in y_from..y_to do
                if words[0] == "toggle"
                    grid[y][x] = !grid[y][x]
                else
                    if words[1] == "on"
                        grid[y][x] = true
                    elsif words[1] == "off"
                        grid[y][x] = false
                    end
                end
            end
        end
    end

    total = 0
    for x in 0..999 do
        for y in 0..999 do
            if grid[y][x]
                total += 1
            end
        end
    end

    return total
end

def part_two(instructions)
    grid = Array.new(1000) {Array.new(1000) {0}}

    instructions.each do | instruction |
        words = instruction.split()

        first, second = words[-3], words[-1]
        x_from, y_from = first.split(",").map(&:to_i)
        x_to, y_to = second.split(",").map(&:to_i)

        lights_affected = (x_to - x_from) * (y_to - y_from)

        for x in x_from..x_to do
            for y in y_from..y_to do
                if words[0] == "toggle"
                    grid[y][x] += 2
                else
                    if words[1] == "on"
                        grid[y][x] += 1
                    elsif words[1] == "off"
                        if grid[y][x] > 0
                            grid[y][x] -= 1
                        end
                    end
                end
            end
        end
    end

    total = 0
    for x in 0..999 do
        for y in 0..999 do
            total += grid[y][x]
        end
    end

    return total
end

instructions = File.readlines("day06.txt")

puts part_one(instructions)
puts part_two(instructions)

File: 2015/Scala/day06.scala
import scala.io.Source
import scala.util.Using

object day06 {

  def partOne(instructions: List[(Int, Array[Int], Array[Int])]): Int = {
    val grid = Array.fill(1000, 1000)(-1) // initialise to -1 (off)

    instructions.foreach { case (state, fromCoords, toCoords) =>
      for x <- fromCoords(0) to toCoords(0)
          y <- fromCoords(1) to toCoords(1) do
        state match
          case  1 => grid(x)(y)  =  1 // on
          case -1 => grid(x)(y)  = -1 // off
          case  0 => grid(x)(y) *= -1 // toggle
    }

    // Return count of how many lights are now on
    grid.flatten.count(_ > 0)
  }

  def partTwo(instructions: List[(Int, Array[Int], Array[Int])]): Int = {
    val grid = Array.fill(1000, 1000)(0) // initialise to 0 brightness

    instructions.foreach { case (state, fromCoords, toCoords) =>
      for x <- fromCoords(0) to toCoords(0)
          y <- fromCoords(1) to toCoords(1) do
        state match
          case  1 => grid(x)(y) += 1 // up
          case -1 => grid(x)(y)  = math.max(grid(x)(y) - 1, 0) // down
          case  0 => grid(x)(y) += 2 // up 2
    }

    // Return total brightness
    grid.flatten.sum
  }

  def main(args: Array[String]): Unit = {
    val inputFile = "2015/data/day06.txt"

    val instructions = Using.resource(Source.fromFile(inputFile)) { source =>
      val splitCoords = (c: String) => c.split(",").map(_.toInt)
      source.getLines().map(line => {
        val words = line.split(" ")
        if words.size == 5 then {
          // turn on/off x,y through x,y
          val state = if words(1).equals("on") then 1 else -1
          (state, splitCoords(words(2)), splitCoords(words(4)))
        } else {
          // toggle x,y through x,y
          (0, splitCoords(words(1)), splitCoords(words(3)))
        }
      }).toList
    }

    println(partOne(instructions))
    println(partTwo(instructions))
  }

}
